javaslang/src/main/java/javaslang/collection/Array.java

```diff
@@ -807,23 +801,25 @@ public int length() {
     public Array<Array<T>> permutations() {
         if (isEmpty()) {
             return empty();
+        } else if (length() == 1) {
+            return of(this);
         } else {
-            final Array<T> tail = tail();
-            if (tail.isEmpty()) {
-                return of(this);
-            } else {
-                final Array<Array<T>> zero = empty();
-                return distinct().foldLeft(zero, (xs, x) -> {
-                    final Function<Array<T>, Array<T>> prepend = l -> l.prepend(x);
-                    return xs.appendAll(remove(x).permutations().map(prepend));
-                });
+            Array<Array<T>> results = empty();
```

Comment:
eliminated recursion inside a foldLeft to make it debuggable and avoid problems with the stack